lib/tree.rb

```diff
@@ -16,47 +16,114 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n)
+  # Space Complexity: O(log n)
   def add(key, value)
-    raise NotImplementedError
+    @root = recursive_add_helper(@root, key, value)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def recursive_add_helper(node, key, value)
+    if node.nil?
+      return TreeNode.new(key, value)
+    elsif key < node.key
+      node.left = recursive_add_helper(node.left, key, value)
+    else
+      node.right = recursive_add_helper(node.right, key, value)
+    end
+    return node
+  end
+
+  # Time Complexity: O(log n)
+  # Space Complexity: O(log n)
   def find(key)
-    raise NotImplementedError
+    find_helper(key, @root)
+  end
+
+  def find_helper(key, current)
+    if current.nil?
+      return nil
+    end
+    if key == current.key
+      return current.value
+    end
+    if current.key > key
+      find_helper(key, current.left)
+    else
+      find_helper(key, current.right)
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def inorder
-    raise NotImplementedError
+    nodes_array = []
+    inorder_helper(@root, nodes_array)
+    return nodes_array
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(node, nodes_array)
+    return if node.nil?
+    inorder_helper(node.left, nodes_array)
+    nodes_array << {:key => node.key, :value => node.value}
+    inorder_helper(node.right, nodes_array)
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def preorder
-    raise NotImplementedError
+    nodes_array = []
+    preorder_helper(@root, nodes_array)
+    return nodes_array
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(node, nodes_array)
+    return if node.nil?
+    nodes_array << {:key => node.key, :value => node.value}
+    preorder_helper(node.left, nodes_array)
+    preorder_helper(node.right, nodes_array)
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def postorder
-    raise NotImplementedError
+    nodes_array = []
+    postorder_helper(@root, nodes_array)
+    return nodes_array
+  end
+
+  def postorder_helper(node, nodes_array)
+    return if node.nil?
+    postorder_helper(node.left, nodes_array)
+    postorder_helper(node.right, nodes_array)
+    nodes_array << {:key => node.key, :value => node.value}
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n)
+  # Space Complexity: O(log n)
   def height
```

Comment:
👍 The time complexities are O(n) since you visit each node.  The space complexity is O(log n) assuming the tree is balanced.